src/circuit/gadget/sinsemilla/chip/hash_to_point.rs

```diff
@@ -0,0 +1,447 @@
+                    .map(|(((x_a, y_a), x_p), y_p)| (y_a - y_p) * (x_a - x_p).invert().unwrap());
+
+                // Assign lambda_1
+                region.assign_advice(
+                    || "lambda_1",
+                    config.lambda_1,
+                    offset + row,
+                    || lambda_1.ok_or(Error::SynthesisError),
+                )?;
+
+                lambda_1
+            };
+
+            // Compute `x_r`
+            let x_r = lambda_1
+                .zip(x_a.value())
+                .zip(x_p)
+                .map(|((lambda_1, x_a), x_p)| lambda_1.square() - x_a - x_p);
+
+            // Compute and assign `lambda_2`
+            let lambda_2 = {
+                let lambda_2 = x_a.value().zip(y_a.0).zip(x_r).zip(lambda_1).map(
+                    |(((x_a, y_a), x_r), lambda_1)| {
+                        pallas::Base::from_u64(2) * y_a * (x_a - x_r).invert().unwrap() - lambda_1
+                    },
+                );
+
+                region.assign_advice(
+                    || "lambda_2",
+                    config.lambda_2,
+                    offset + row,
+                    || lambda_2.ok_or(Error::SynthesisError),
+                )?;
+
+                lambda_2
+            };
+
+            // Compute and assign `x_a` for the next row.
+            let x_a_new: X<pallas::Base> = {
+                let x_a_new = lambda_2
+                    .zip(x_a.value())
+                    .zip(x_r)
+                    .map(|((lambda_2, x_a), x_r)| lambda_2.square() - x_a - x_r);
+
+                let x_a_cell = region.assign_advice(
+                    || "x_a",
+                    config.x_a,
+                    offset + row + 1,
+                    || x_a_new.ok_or(Error::SynthesisError),
+                )?;
+
+                CellValue::new(x_a_cell, x_a_new).into()
+            };
+
+            // Compute y_a for the next row.
+            let y_a_new: Y<pallas::Base> = lambda_2
+                .zip(x_a.value())
+                .zip(x_a_new.value())
+                .zip(y_a.0)
+                .map(|(((lambda_2, x_a), x_a_new), y_a)| lambda_2 * (x_a - x_a_new) - y_a)
+                .into();
+
+            // Update the mutable `x_a`, `y_a` variables.
+            x_a = x_a_new;
+            y_a = y_a_new;
+        }
+
+        Ok((x_a, y_a, zs))
+    }
+}
+
+// The x-coordinate of the accumulator in a Sinsemilla hash instance.
+struct X<F: FieldExt>(CellValue<F>);
+
+impl<F: FieldExt> From<CellValue<F>> for X<F> {
+    fn from(cell_value: CellValue<F>) -> Self {
+        X(cell_value)
+    }
+}
+
+impl<F: FieldExt> Deref for X<F> {
+    type Target = CellValue<F>;
+
+    fn deref(&self) -> &CellValue<F> {
+        &self.0
+    }
+}
+
+// The y-coordinate of the accumulator in a Sinsemilla hash instance.
+// This is never actually witnessed until the last round, since it
+// can be derived from other variables. Thus it only exists as a field
+// element, not a `CellValue`.
```

Comment:
```suggestion
/// The y-coordinate of the accumulator in a Sinsemilla hash instance.
///
/// This is never actually witnessed until the last round, since it
/// can be derived from other variables. Thus it only exists as a field
/// element, not a `CellValue`.
```